### PR TITLE
Avoid parent metadata reference for local components

### PR DESCRIPTION
**Pull request by jvandort**
We previously relied on the parent component when calculating configuration artifacts in order to get the artifacts of other configurations in the component. We should deprecate this behavior of inheriting artifacts, but in the meantime, we avoid the parent reference by re-calculating the artifact instances when we do inherit artifacts.

This is a bit inefficient, but should not affect performance. As a result, we get to avoid the ugly circular parent reference but also get to avoid the `getConfiguration` call on the component. We want to remove this method, as we want to get rid of the configuration concept from the resolution engine.

Furthermore, two additional commits are included that 
1. remove a double-layered `Optional<List<>>` in an interface, as we treated the empty optional the same as an empty list, and this double nested generics made it impossible for sub-interfaces to refine the type in the list
2. Rename some methods 

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

### CODE REVIEW

**Comment by jvandort** [comment]
@bot-gradle test this

**Comment by jvandort** [comment]
@bot-gradle test this

**Comment by jvandort** [comment]
@bot-gradle test this

File: platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/moduleconverter/dependencies/DefaultLocalConfigurationMetadataBuilder.java [hunk 1]
```diff
@@ -151,33 +140,31 @@ public void visitChildVariant(String name, DisplayName displayName, ImmutableAtt
     }
 
     private static CalculatedValue<ImmutableList<LocalComponentArtifactMetadata>> getConfigurationArtifacts(
-        LocalComponentMetadata parent, ModelContainer<?> model, CalculatedValueContainerFactory calculatedValueContainerFactory,
-        String name, String description, ImmutableSet<String> hierarchy, List<PublishArtifact> sourceArtifacts
+        ModelContainer<?> model, CalculatedValueContainerFactory calculatedValueContainerFactory,
+        String name, String description, ImmutableSet<String> hierarchy, ConfigurationsProvider configurationsProvider, ComponentIdentifier componentId
```
Comment (big-guy) [inline platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/moduleconverter/dependencies/DefaultLocalConfigurationMetadataBuilder.java:144]: 💅 could you maybe reorder these parameters to group them by "type" (services vs model vs inputs?)

**Review by big-guy — APPROVED** [review]

**Pull request merged**